vm/concurrent_hash.go

```diff
@@ -0,0 +1,377 @@
+package vm
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/goby-lang/goby/vm/classes"
+	"github.com/goby-lang/goby/vm/errors"
+)
+
+// ConcurrentHash is a thread-safe implementation of a Hash.
+//
+// Due to the implementation (based on sync.Map), iteration is not deterministic under all
+// circumstances (see https://golang.org/pkg/sync/#Map).
+//
+// ```ruby
+// a = ConcurrentHash.new({ balthazar1: 100 })
+//
+// a["balthazar1"]  # => 100
+// ```
+//
+// - **value:** String literal and objects (Integer, String, Array, Hash, nil, etc) can be used.
+//
+// **Note:**
+// - The order of key-value pairs are **not** preserved.
+type ConcurrentHashObject struct {
+	*baseObj
+	internalMap sync.Map
+}
+
+// Class methods --------------------------------------------------------
+func builtinConcurrentHashClassMethods() []*BuiltinMethodObject {
+	return []*BuiltinMethodObject{
+		{
+			Name: "new",
+			Fn: func(receiver Object, instruction *instruction) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *normalCallFrame) Object {
+					if len(args) > 1 {
+						return t.vm.initErrorObject(errors.ArgumentError, instruction, "Expect 0 or 1 arguments, got %d", len(args))
+					}
+
+					if len(args) == 0 {
+						return t.vm.initConcurrentHashObject(make(map[string]Object))
+					} else {
+						arg := args[0]
+						hashArg, ok := arg.(*HashObject)
+
+						if !ok {
+							return t.vm.initErrorObject(errors.TypeError, instruction, errors.WrongArgumentTypeFormat, classes.HashClass, arg.Class().Name)
+						}
+
+						return t.vm.initConcurrentHashObject(hashArg.Pairs)
+					}
+				}
+			},
+		},
+	}
+}
+
+// Instance methods -----------------------------------------------------
+func builtinConcurrentHashInstanceMethods() []*BuiltinMethodObject {
```

Comment:
@saveriomiroddi  Are these methods' implementation same as the original Hash? If so, how about make

```go
h := receiver.(*ConcurrentHashObject)
// or
h := receiver.(*HashObject)
```
become something like:

```go
type hash type {
    load() Object, bool
    store(Object)
}

h := receiver.(hash)
v, ok := h.load()
```

Then we can just implement the interface on both `HashObject` and `ConcurrentHashObject`. After that we can use same method implementations for both classes. (This could be a large refactoring though)

Reply:
`sync.Map` has some API and semantic differences; the generic associative array functionality can be abstracted, but not completely (at least, not completely and exactly).

Have a look at the `sync.Map` documentation and let me know, since some design decisions need to be taken, in particular, due the non-determinism of the iteration.

Reply:
I know `sync.Map` has its own API, so what I suggest is to wrap another layer around both normal map and `sync.Map`'s APIs. For example:

```go
func (*ch ConcurrentHashObject) load(key *StringObject) Object {
    value, ok := h.internalMap.Load(key.value)
    if !ok {
        return NULL
    }
   return value.(Object)
}

func (*h HashObject) load(key *StringObject) Object {
    value, ok :=h.Pairs[key.value]
    if !ok {
        return NULL
    }
   return value
}
```

Then if we implement all the Hash methods using those interface APIs, we can share those methods between these two types.

Reply:
The problems I refer to are:

1. iteration: supported with different API
2. length: it's not supported at all by sync.Map

Problem 1. can be solved easily, although a bit clunkily, by abstracting a `range()` function, and having Hash methods pass a function to it, instead of using the built-in `range` keyword

Problem 2. is significant, as it has both functional and semantic implications. the functional problem can be solved (if necessary) through count via iteration (which will be in linear rather than constant time) . the semantic aspect is something that needs to be explicitly accepted or rejected; if accepted, it needs to be documented very clearly that iteration (and count) will be non-deterministic.

Reply:
I think I'll drop the `length` method from concurrent map.

I just took a look at [this issue](https://github.com/golang/go/issues/20680) and there's no clear usage for the `Len` method of `sync.Map`. So I assume there won't be much usage for concurrent hash's `length` method either. And user can use iteration to count the number themselves if needed.

Reply:
:+1: